### PR TITLE
add plugin that configures centos7 nodes in preparation to have rke r…

### PR DESCRIPTION
**Pull request by danieltmiles**
…un on them [DEV-12907]

[DEV-12907]: https://cloudbolt.atlassian.net/browse/DEV-12907

### CODE REVIEW

**Comment by danieltmiles** [comment]
From Bernard in a meeting, we could do with some explanatory comments and maybe some long-method-smell fighting.

**Comment by danieltmiles** [comment]
convert print statements to logger statemwents

**Comment by danieltmiles** [comment]
there is a vardir in cloudbolt settings, use that instead of hard-coding /var

File: actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+                    'groupadd docker\n' + \
+                    'usermod -aG docker {} || exit 1\n'.format(user) + \
+                    'mkdir -p /home/{}/.ssh || exit 1;\n'.format(user) + \
+                    'echo \'{}\' >> /home/{}/.ssh/authorized_keys || exit 1;\n'.format(ssh_public_key, user) + \
+                    'chown -R {} /home/{}/.ssh || exit 1;\n'.format(user, user) + \
+                    'chmod 755 /home/{}/.ssh || exit 1;\n'.format(user) + \
+                    'chmod 644 /home/{}/.ssh/authorized_keys || exit 1;\n'.format(user) + \
+                    'echo \'net.ipv6.conf.all.forwarding=1\' >> /etc/sysctl.conf || exit 1\n' + \
+                    'sysctl -p /etc/sysctl.conf || exit 1'
+    for tcp_port_num in [443, 10250, 2379, 2380, 6443]:
```
Comment (ignacious) [inline actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py]: Add port 80

File: actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    docker_script += '\nsystemctl restart firewalld;\n'
+    logger.info(f"docker script:\n{docker_script}")
+    for server in find_all_servers(blueprint_context=blueprint_context):
+        set_progress("Starting script execution on server {}".format(server.ip))
+        server.execute_script(script_contents=docker_script, timeout=700)
+        server.reboot()
+    set_progress("Waiting for server(s) to begin reboot.")
+    time.sleep(10)
+    for server in find_all_servers(blueprint_context=blueprint_context):
+        server.wait_for_os_readiness()
```
Comment (ignacious) [inline actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py]: wait_for_os_readiness may be vcenter specific - particularly the CB RH

File: actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+                    'useradd {};\n'.format(user) + \
+                    'groupadd docker\n' + \
+                    'usermod -aG docker {} || exit 1\n'.format(user) + \
+                    'mkdir -p /home/{}/.ssh || exit 1;\n'.format(user) + \
+                    'echo \'{}\' >> /home/{}/.ssh/authorized_keys || exit 1;\n'.format(ssh_public_key, user) + \
+                    'chown -R {} /home/{}/.ssh || exit 1;\n'.format(user, user) + \
+                    'chmod 755 /home/{}/.ssh || exit 1;\n'.format(user) + \
+                    'chmod 644 /home/{}/.ssh/authorized_keys || exit 1;\n'.format(user) + \
+                    'echo \'net.ipv6.conf.all.forwarding=1\' >> /etc/sysctl.conf || exit 1\n' + \
+                    'sysctl -p /etc/sysctl.conf || exit 1'
```
Comment (pop) [inline actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py]: I'd propose, if we're going to put Bash inline, that we do it like this:
```py
docker_script = [
    "yum -y install docker || exit 1;",
    "systemctl enable docker || exit 1;",
    "useradd {username};",
    "groupadd docker;",
    "usermod -aG docker {username} || exit 1;",
    "mkdir -p /home/{username}/.ssh || exit 1;",
    "echo '{pub_key}' >> /home/{username}/.ssh/authorized_keys || exit 1;",
    "chown -R {username} /home/{username}/.ssh || exit 1;",
    "chmod 755 /home/{username}/.ssh || exit 1;",
    "'chmod 644 /home/{username}/.ssh/authorized_keys || exit 1;",
    "echo 'net.ipv6.conf.all.forwarding=1' >> /etc/sysctl.conf || exit 1;",
    "sysctl -p /etc/sysctl.conf || exit 1;",
].join("\n").format(
    username=user,
    pub_key=ssh_public_key,
)
```
it's a little cleaner and less error prone for refactors.

File: actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py (hunk 1 above)
Comment (pop) [reply actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py]: Let's make that info a comment.

Also should we be exposing 80? We aren't right now.

File: actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    docker_script += '\nsystemctl restart firewalld;\n'
+    logger.info(f"docker script:\n{docker_script}")
+    for server in find_all_servers(blueprint_context=blueprint_context):
+        set_progress("Starting script execution on server {}".format(server.ip))
+        server.execute_script(script_contents=docker_script, timeout=700)
+        server.reboot()
+    set_progress("Waiting for server(s) to begin reboot.")
+    time.sleep(10)
+    for server in find_all_servers(blueprint_context=blueprint_context):
+        server.wait_for_os_readiness()
```
Comment (pop) [inline actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py]: Let's add some veritcal to this script. It'll make it much more readable.

File: actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    ssh_public_key, ssh_private_key = create_ssh_keypair()
+    prepare_server_hosts(user, blueprint_context, ssh_public_key)
+
+    os.makedirs(RKEDIR, exist_ok=True)
+    with open(f"{RKEDIR}/rke_private_key.pem", 'w') as fl:
+        fl.write(ssh_private_key)
+    with open(f"{RKEDIR}/rke_public_key.pem", 'w') as fl:
+        fl.write(ssh_public_key)
+
+    ips = find_all_server_ips(kwargs.get('blueprint_context', {}))
```
Comment (pop) [inline actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py]:
```suggestion
    ips = find_all_server_ips(blueprint_context)
```
I _think_ we can make that change.

File: actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    with open(f"{RKEDIR}/{cluster_yml_name}", 'w') as fl:
+        fl.write(rke_yaml_text)
+
+    set_progress(f"Your ssh public and private keys have been generated. Please find them in {RKEDIR}")
+    set_progress(f"Your RKE cluster.yml file has been generated. Please find it in {RKEDIR}/{cluster_yml_name}")
+
+    set_progress(f"Running RKE Config")
+    kubernetes_up(cluster_yml_name)
+
+    return "SUCCESS", f"./rke up --config={RKEDIR}/{cluster_yml_name}", ""
```
Comment (pop) [inline actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py]: I forget, is there a reason we're not running this in the script?

File: actions/cloudbolt_plugins/configure_nodes_for_rke/configure_nodes_for_rke.json [hunk 7]
```diff
@@ -0,0 +1,10 @@
+{
+    "description": "",
+    "max-retries": 0,
+    "name": "configure-nodes-for-rke",
+    "resource-technologies": [],
+    "script-filename": "cb_plugin_1570640510857138_mjMOkye.py",
```
Comment (pop) [inline actions/cloudbolt_plugins/configure_nodes_for_rke/configure_nodes_for_rke.json]: We can rename the script something more human readable with something like `git mv cb_plugin_1570640510857138_mjMOkye.py configure_nodes_for_rke.py`

File: actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py (hunk 1 above)
Comment (ignacious) [reply actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py]: good catch. yeah Dan mentioned we should add that in.

File: actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py (hunk 6 above)
Comment (ignacious) [reply actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py]: As i understand it, i think we were running this script per server. The cluster file should work for multiple servers though. https://github.com/CloudBoltSoftware/cloudbolt-forge/pull/38/files/57aed45df149fa729725e234434f8149d3715b6e#diff-7e89c850b9a7377b9415eaf1d5313401R45

File: actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py (hunk 5 above)
Comment (pop) [reply actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py]: Any word on this change?

File: actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py (hunk 3 above)
Comment (ignacious) [reply actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py]: I did attempt this but ran into some issues in implementation. I'd like to revisit this at a later date.

File: actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py (hunk 5 above)
Comment (ignacious) [reply actions/cloudbolt_plugins/configure_nodes_for_rke/cb_plugin_1570640510857138_mjMOkye.py]: I tried this and exception was thrown. will need more investigation.

File: actions/cloudbolt_plugins/configure_nodes_for_rke/README.md [hunk 8]
```diff
@@ -0,0 +1,6 @@
+This plugin will build and deploy a vanilla kubernetes cluster to each server supplied in the blueprint.
+
+You will need to manually install rke. See https://github.com/rancher/rke/releases. This plugin assumes an rke executable in the '/var/opt/cloudbolt/rke/' path.
```
[the remark this answers is not in the document]
Comment (ignacious) [reply actions/cloudbolt_plugins/configure_nodes_for_rke/README.md]: Agreed.

File: actions/cloudbolt_plugins/configure_nodes_for_rke/README.md [hunk 9]
```diff
@@ -0,0 +1,6 @@
+This plugin will build and deploy a vanilla kubernetes cluster onto the servers provisioned in a blueprint. Each server will be assigned one or more roles from: API server, worker, and etcd.
```
Comment (jevanscb) [inline actions/cloudbolt_plugins/configure_nodes_for_rke/README.md]:
```suggestion
This plugin will build and deploy a vanilla Kubernetes cluster onto the servers provisioned in a Blueprint. Each server will be assigned one or more roles from: API server, worker, and etcd.
```

File: actions/cloudbolt_plugins/configure_nodes_for_rke/README.md [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+
+You will need to manually install `rke` to your CloudBolt server. See https://github.com/rancher/rke/releases for insructions. This plugin assumes an `rke` executable exists in the '/var/opt/cloudbolt/rke/bin/' directory.
+
+This plugin allows for multiple clusters. Configuration for each cluster is based off of the resource associated with the deployed blueprint.
```
Comment (jevanscb) [inline actions/cloudbolt_plugins/configure_nodes_for_rke/README.md]:
```suggestion
This plugin allows for multiple clusters. Configuration for each cluster is based off of the resource associated with the deployed Blueprint.
```

File: actions/cloudbolt_plugins/configure_nodes_for_rke/README.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+> !! CENTOS ONLY !! This plugin only works for CentOS servers and requires the `yum` package manager.
+
+This plugin builds and deploys a vanilla Kubernetes cluster onto the servers provisioned in a Blueprint.
+Each server is assigned one or more roles from: Kube API server, Kube worker, and etcd server.
+
+## Usage
+
+Create a Blueprint with this plugin as the step _after_ one or more servers have been provisioned.
+
+After the blueprint order completes you can grab the Kubernetes cluster information and create a CloudBolt Container Orchestrator.
```
Comment (jevanscb) [inline actions/cloudbolt_plugins/configure_nodes_for_rke/README.md]:
```suggestion
After the Blueprint order completes, you can grab the Kubernetes cluster information and create a CloudBolt Container Orchestrator.
```

File: actions/cloudbolt_plugins/configure_nodes_for_rke/README.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+It also requires SSH from the CloudBolt server to the newly created servers, so you should expose these ports too:
+
+* 22/tcp
+
+If you are running in the cloud, add the above ports to a security group allowing incoming traffic from IP ranges including the CloudBolt server and the newly provisioned servers.
+
+### Node size
+
+Note that Kubernetes clusters, especially those with only a few nodes, have a lot of overhead in CPU and memory.
```
Comment (jevanscb) [inline actions/cloudbolt_plugins/configure_nodes_for_rke/README.md]:
```suggestion
Note that Kubernetes clusters, especially those with only a few nodes, have a lot of overhead in CPU and memory, 
```

File: actions/cloudbolt_plugins/configure_nodes_for_rke/README.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+So using smaller node sizes may be slow and error/timeout-prone.
```
Comment (jevanscb) [inline actions/cloudbolt_plugins/configure_nodes_for_rke/README.md]:
```suggestion
so using smaller node sizes may be slow and error/timeout-prone.
```

File: actions/cloudbolt_plugins/configure_nodes_for_rke/README.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+We suggest using at least "medium" size nodes with this Plugin to avoid that overhead performance cost.
+
+## Deployment Specifics
+
+This plugin will provision one Kubernetes cluster per Catalog Item order.
+
+Each cluster is unique to the Resource and the clusters have files on disk which can be inspected for troubleshooting purposes.
+For example a Resource with ID 101 can be found on the CloudBolt host at `/var/opt/cloudbolt/rke/resource-101`.
```
Comment (jevanscb) [inline actions/cloudbolt_plugins/configure_nodes_for_rke/README.md]:
```suggestion
For example, a Resource with ID 101 can be found on the CloudBolt host at `/var/opt/cloudbolt/rke/resource-101`.
```

**Pull request merged**